client/my-sites/plugins/with-plugin-redirect.js

```diff
@@ -0,0 +1,150 @@
+
+// Plugins list that should perfomr a redirect
+// once it's installed.
+const redirectingPluginsList = {
+	woocommerce: getSiteWoocommerceWizardUrl,
+};
+
+/**
+ * Return the handler function to get the plugin redirect URL.
+ *
+ * @param {string} slug - Plugin slug.
+ * @returns {Function|boolean} Selector handler if the plugin should redirect. Otherwise, False.
+ */
+function getPluginRedirectUrl( slug ) {
+	if ( Object.keys( redirectingPluginsList ).indexOf( slug ) < 0 ) {
+		return false;
+	}
+
+	return redirectingPluginsList[ slug ];
+}
+
+const withPluginRedirect = ( Component ) => ( props ) => {
+	const { plugin = {} } = props;
+	const pluginSlug = plugin?.slug;
+
+	if ( ! pluginSlug ) {
+		return <Component { ...props } />;
+	}
+
+	const redirectUrlHandler = getPluginRedirectUrl( pluginSlug );
+	if ( ! redirectUrlHandler ) {
+		return <Component { ...props } />;
+	}
+
+	const translate = useTranslate();
+	const dispatch = useDispatch();
+
+	const { redirect, url, hasSiteBeenTransferred } = useSelector(
+		( state ) => {
+			const site = getSelectedSite( state );
+			const isAtomic = isSiteAutomatedTransfer( state, site.ID );
+			const isJetpack = isJetpackSite( state, site.ID );
+			const woocommerceWizardUrl = redirectUrlHandler( state, site.ID );
+			const transferState = getAutomatedTransferStatus( state, site.ID );
+			const hasPluginJustBeenInstalled = isPluginActionCompleted(
+				state,
+				site.ID,
+				pluginSlug,
+				INSTALL_PLUGIN
+			);
+
+			const isSiteConnected = getSiteConnectionStatus( state, site.ID );
+			const selectedData = {
+				redirect: false,
+				url: woocommerceWizardUrl,
+				hasSiteBeenTransferred: transferState === transferStates?.COMPLETE,
+			};
+
+			// Jetpack site.
+			if (
+				isJetpack &&
+				pluginSlug === 'woocommerce' &&
+				woocommerceWizardUrl &&
+				hasPluginJustBeenInstalled &&
+				isSiteConnected
+			) {
+				return { ...selectedData, redirect: true };
+			}
+
+			// Atomic Site.
+			if (
+				isAtomic &&
+				pluginSlug === 'woocommerce' &&
+				woocommerceWizardUrl &&
+				selectedData.hasSiteBeenTransferred &&
+				isSiteConnected
+			) {
+				return { ...selectedData, redirect: true };
+			}
+
+			return selectedData;
+		},
+		[ pluginSlug ]
+	);
+
+	const pluginName = plugin?.name;
+
+	useEffect( () => {
+		if ( ! redirect || ! url || ! pluginName ) {
+			return;
+		}
+
+		const pluginInstalledNoticeTime = hasSiteBeenTransferred ? 0 : 3000;
+
+		let timerId = setTimeout( () => {
+			dispatch( removeNotice( 'plugin-notice' ) );
+
+			// re-use same timer ID to redirect the client.
+			timerId = setTimeout( () => {
+				window.location.href = url;
+			}, 5500 );
+
+			dispatch(
+				successNotice(
+					translate( 'Redirecting to the %(pluginName)s onborading page in five seconds.', {
+						args: {
+							pluginName,
+						},
+					} ),
+					{
+						duration: 5000,
+						button: translate( 'Cancel' ),
+						id: 'plugin-redirect-notice',
+						onClick: function () {
```

Comment:
I'm wondering about accessibility here. Can it be cancelled using the escape key for example? Should we also have a 'redirect now' link, or make part of the text a link, so you don't have to wait? With that, we could extend the delay too.

Reply:
I can dig into this, but I'm not sure whether is possible to add another button there. Also, we could do the opposite by providing a button to redirect instead.

Reply:
> This is working well for me. The redirect happens after a few seconds. I didn't end up in exactly the right place at one point, but I believe that was related to the setup of a support session interfering with the redirect.

what was the url?


> I noticed that if I navigate away from the page, when I return for the first time, I'm prompted that the redirect will happen in 5 seconds. I think this is a good reason to have the delay, and not redirect immediately.

In theory, when the redirection should be canceled when the PluginMeta is unMounted. Let me check it.


> I'm not sure if we should tackle it with this PR if it's a problem, but I think the routing isn't working correctly. If I go to `/plugins/woocommerce` without a site slug, I'm not getting the site selector.

Weird. Not related to this PR if I'm not wrong, though.